pdns/common_startup.cc

```diff
@@ -673,12 +675,36 @@ void mainthread()
     sd_notify(0, "READY=1");
 #endif
 
+  const uint32_t secpollInterval = 1800;
+  uint32_t secpollSince = 0;
+  uint32_t domainCacheUpdateSince = 0;
   for(;;) {
-    sleep(1800);
-    try {
-      doSecPoll(false);
+    const uint32_t slept = g_domainCache.getTTL() == 0 ? secpollInterval : std::min(secpollInterval, g_domainCache.getTTL());
+    sleep(slept);  // if any signals arrive, we might run more often than expected.
```

Comment:
This is a very confusing line, because `sleep` -returns- the number of seconds slept, but we don't use that value (and the 'might run more often' comment also suggests you actually wanted the return value!).

Maybe rename your `slept` to `sleeptime`?